delta-lake/delta-22x/src/main/scala/com/nvidia/spark/rapids/delta/DeltaLogShim.scala

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.delta.shims
+
+import org.apache.spark.sql.delta.DeltaLog
+import org.apache.spark.sql.delta.actions.Metadata
+import org.apache.spark.sql.execution.datasources.FileFormat
+
+object DeltaLogShim {
+  def fileFormat(deltaLog: DeltaLog): FileFormat = {
+    deltaLog.fileFormat(deltaLog.unsafeVolatileSnapshot.metadata)
+  }
+  def getMetadata(deltaLog: DeltaLog): Metadata = {
+    deltaLog.unsafeVolatileSnapshot.metadata
+  }
+}
```

Comment:
This shim is problematic because it conflicts with the delta-21x implementation when the jar is built.  spark330 depends on both delta-21x and delta-22x, but only one copy of the com.nvidia.spark.rapids.delta.shims.DeltaLogShim class can exists in the spark330 parallel-world layout.

That's why every file under the delta-lake/delta-2* projects is in a package specific to that project (e.g.: com.nvidia.spark.rapids.delta.delta21x vs com.nvidia.spark.rapids.delta.delta22x).  They never use a common package that appears across projects so they cannot conflict when combined under a single spark version's parallel-world directory.

To work around this issue, these shims are implemented via DeltaRuntimeShim which uses a different package name per shim so they can coexist in the same spark parallel-world layout.  There's already an existing shim for unsafeVolatileSnapshot which can get the snapshot and the metadata should be commonly accessible from the snapshot.  Suggest adding a fileFormatFromLog method to DeltaRuntimeShim and handle it that way.

Compile-time shims for Delta in the Databricks area is OK, because we never support multiple Delta Lake versions per Spark version there.  It's only an issue in delta-io because multiple Delta Lake versions can map to a single Spark version.

Reply:
This is now fixed.